README.md
# mb.extruder 1.6

__An open source jQuery component to easily build a top or left sliding panel.__

![mb.extruder](http://pupunzi.com/gitHub/mb.extruder.jpg)

## [go to the demo](http://pupunzi.com/#mb.components/mb.extruder/extruder.html)
## [go to the doc](http://wiki.github.com/pupunzi/jquery.mb.extruder/)
## [go to the blog](http://pupunzi.open-lab.com/mb-jquery-components/jquery-mb-extruder/)


[jquery.mb.components](http://pupunzi.com/), another way of thinking the web


inc/changeLog.txt
jquery.mb.extruder


2.2
    1. added "autoOpenTime" to control the open on mouseover behaviour. (0= do not open onmouseover - value in milliseconds)

2.1
    1. Solved an issue with jquery 1.4.3. (.content display none was not resetted once opened with animate and jquery 1.4.3)

2.0
    1. extruder on top, left, right, bottom.
    2. added a new property: accordionPanels to choose if panels shold close once opening a new one or not.

1.8.9
    1. bugfix IE.
    2. added extruder right

1.8.5
    1. Add close extruder on flap click.

1.8.1
    1. bugfix for Webkit browsers: the flap didn't align well.

1.8.0
    1. Extruder is child of any DOM element.
    2. Extruder can have inline content. (not only Ajax).
    3. better close event.
    4. added autoCloseTime parameter to specify time in milliseconds the extruder should close on mouse out (0 = never).
    5. added hidePanelsOnClose parameter to specify if subpanels must should hide on extruder close.

1.6
    1. Removed the ".content" div from HTML
    2. support content defined in HTML

1.5
  modified the flipText behavior.
  added "positionFixed" param to control the position of the extruder (true id default and the extruder doesn't scroll with the page; if set to false it scrolls).
  added a method to disable/enable a voice on the fly: $.fn.disableExtruderVoice and $.fn.enableExtruderVoice.

1.0
  Initial release.

inc/mbExtruder.pack.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: info@pupunzi.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * jQuery.mb.components: mb.extruder
 * version: 1.0 - 25-ott-2009 - 32
 * © 2001 - 2009 Matteo Bicocchi (pupunzi), Open Lab
 *
 */

eval(function(p,a,c,k,e,r){e=function(c){return(c<a?'':e(parseInt(c/a)))+((c=c%a)>35?String.fromCharCode(c+29):c.toString(36))};if(!''.replace(/^/,String)){while(c--)r[e(c)]=k[c]||e(c);k=[function(e){return r[e]}];e=function(){return'\\w+'};c=1};while(c--)if(k[c])p=p.replace(new RegExp('\\b'+e(c)+'\\b','g'),k[c]);return p}('(7($){G.13=0;G.14=0;i g=$.1W.1X;$.H={1Y:"1Z 20",21:"1.5",1q:{n:22,1r:15,1s:23,m:"s",16:1,1t:D,1u:25,1v:"24",N:7(){},17:7(){},O:7(){},P:26},18:7(f){I 3.19(7(){3.6={};$.1a(3.6,$.H.1q);$.1a(3.6,f);i a,V,1b,1c,J,1d;a=$(3);a.h("W",D);i b=3.6.1r?"1w":"X";9(3.6.m=="s"){G.14++;9(G.14>1){1x("27 28 1 29.1y 2a s 2b 1e 2c 2d... 2e 2f!");I}}a.z("1y");a.z(3.6.m);1b=3.6.m=="s"?{m:b,s:0,A:"2g%",2h:-3.6.n/2,n:3.6.n}:{m:b,s:0,A:-(3.6.n)};a.h(1b);9(!g)a.h({o:3.6.16});a.2i("<p t=\'1z\'></p>");V=a.8(".1z");1c={n:3.6.n,m:"X"};V.h(1c);9($.l){$.l.1A("t");9(a.l().1B)a.j("1f",a.l().1B);9(a.l().u)a.j("1g",a.l().u);9(a.l().q)a.j("1C",a.l().q)}V.1h("<p t=\'2j\'></p><p t=\'Q\'><1D t=\'E\'/></p>");a.8(\'.R\').2k("<p t=\'1i\'></p>");J=a.j("1f")?a.j("1f"):"";a.8(".E").2l(J);9(3.6.m=="A"){a.8(".1i").Y(J);a.8(".E").Y(J).h({1E:"1F"});i c=3.6.1v=="2m";i d=a.8(\'.E\').2n()[1];a.8(\'.E\').2o(c)}Z{a.8(".E").Y(J).h({1E:"1F",n:3.6.1t})}9(a.j("1g")){a.11({u:a.j("1g"),q:a.j("1C"),1j:7(){9(a.k(0).6.17)a.k(0).6.17()}})}Z{a.1k()}a.8(\'.Q .E\').2p({2q:7(){a.1l();a.S()},2r:7(){},2s:2,2t:3.6.1s});a.8(\'.Q\').12("B",7(){a.S()});a.8(\'.R\').12("2u",7(){1d=1G(7(){a.K()},2v)}).12("2w",7(){2x(1d)});9(3.6.m=="A"){a.8(\'.R\').h({n:3.6.n,1H:"D%"});a.8(\'.Q\').h({2y:-2z,s:D+G.13});G.13+=d+3.6.1u;i e=$("<p/>").h({m:"X",s:0,A:0,n:"D%",1H:"D%",2A:"2B"});a.8(\'.Q\').1h(e)}})},11:7(b){3.6={u:T,q:"",1j:7(){}};$.1a(3.6,b);9(!3.6.u||3.6.u.1I==0){1x("2C 2D: 2E 2F 2G 2H");I}i c=3.6.u;i d=3.6.q;i e=$(3),L;i f=3.6.1j;$.1J({1K:"1L",u:c,q:d,1M:7(a){e.8(".1i").1N(a);L=e.8(".L");L.F(7(){$(3).z("F")},7(){$(3).v("F")});e.1k();9(f){1G(7(){f()},D)}}})},S:7(){i a=$(3);a.z("1O");9(!g)a.h("o",1);i b=a.k(0).6.m;a.1l();9(b=="s"){a.8(\'.R\').1P(a.k(0).6.P);9(a.k(0).6.N)a.k(0).6.N()}Z{9(!g)$(3).h("o",1);a.1Q({A:0},a.k(0).6.P);9(a.k(0).6.N)a.k(0).6.N()}},K:7(){i a=$(3);a.v("1O");9(!g)a.h("o",a.k(0).6.16);a.U();9(a.k(0).6.m=="s"){a.8(\'.R\').1m(a.k(0).6.P);9(a.k(0).6.O)a.k(0).6.O()}Z 9(a.k(0).6.m=="A"){a.1Q({A:-a.k(0).6.n},a.k(0).6.P,7(){9(a.k(0).6.O)a.k(0).6.O()})}}};1R.w.1l=7(){i b=10;$(\'*\').19(7(){9($(3).h("m")=="X"||$(3).h("m")=="1w"){i a=1S($(3).h(\'W\'));b=a>b?1S($(3).h(\'W\')):b}});$(3).h(\'W\',b+=1);I b};$.w.1k=7(){i e=$(3);i f=$(3).8(".L");f.19(7(){i d=$(3);9($.l){$.l.1A("t");9(d.l().x)d.j("x",d.l().x);9(d.l().q)d.j("q",d.l().q);9(d.l().r)d.j("r",d.l().r)}9(d.j("r")){d.1T()}9(d.j("x")&&d.j("x")!="T"){d.1h("<1D t=\'y\'/>");d.8(".y").h({o:.5});d.8(".y").F(7(){$(3).h({o:1})},7(){$(3).1e(".C").h({o:.5})}).B(7(){9($(3).2I().2J("C")){e.U();I}e.8(".1n").1m(1o,7(){$(3).1U()});f.v("C");f.8(".y").v("C").h({o:.5});i b=$("<p t=\'1n\'></p>");$.1J({1K:"1L",u:d.j("x"),q:d.j("q"),1M:7(a){i c=$(a);b.Y(c);b.2K().z("2L").B(7(){e.U();e.K()});b.1P(1o)}});d.1N(b);d.z("C");d.8(".y").z("C").h({o:1})})}9(d.8("a").1I==0&&d.j("x")){d.8(".M").1e(".r").h("1p","1V").B(7(){d.8(".y").B()})}9((!d.j("x")||d.j("x")=="T")&&(!d.j("r")||d.j("r")!="15")){d.8(".M").B(7(){e.U();e.K()})}})};$.w.1T=7(){i b=$(3);b.j("r",15);b.8(".M").h("o",.4);b.F(7(){$(3).v("F")},7(){$(3).v("F")});b.8(".M").z("r").h("1p","2M");b.8(".y").2N();b.12("B",7(a){a.2O();I T})};$.w.2P=7(){i a=$(3);a.j("r",T);a.8(".M").h("o",1);a.8(".M").v("r").h("1p","1V");a.8(".y").2Q();a.2R("B")};$.w.U=7(){i a=$(3).8(".L");$(3).8(".1n").1m(1o,7(){$(3).1U()});a.v("C");a.8(".y").v("C").h("o",.5)};$.w.18=$.H.18;$.w.11=$.H.11;$.w.K=$.H.K;$.w.S=$.H.S})(1R);',62,178,'|||this|||options|function|find|if||||||||css|var|attr|get|metadata|position|width|opacity|div|data|disabled|top|class|url|removeClass|fn|panel|settingsBtn|addClass|left|click|sel|100|flapLabel|hover|document|mbExtruder|return|txt|closeMbExtruder|voice|label|onExtOpen|onExtClose|slideTimer|flap|content|openMbExtruder|false|closeAllPanel|wrapper|zIndex|absolute|html|else||setMbExtruderContent|bind|extruderLeft|extruderTop|true|extruderOpacity|onExtContentLoad|buildMbExtruder|each|extend|extruderStyle|wrapperStyle|timer|not|extTitle|extUrl|append|header|callback|setExtruderVoicesAction|mb_bringToFront|slideUp|optionsPanel|400|cursor|defaults|positionFixed|sensibility|flapDim|flapMargin|textOrientation|fixed|alert|extruder|ext_wrapper|setType|title|extData|span|whiteSpace|noWrap|setTimeout|height|length|ajax|type|POST|success|after|open|slideDown|animate|jQuery|parseInt|disableExtruderVoice|remove|pointer|browser|msie|author|Matteo|Bicocchi|version|350|800|bt||300|more|than|mb|on|is|supported|jet|hope|soon|50|marginLeft|wrapInner|footer|prepend|text|tb|getFlipTextDim|mbFlipText|hoverIntent|over|out|sensitivity|interval|mouseleave|1000|mouseenter|clearTimeout|marginRight|40|background|transparent|internal|error|no|URL|to|call|parents|hasClass|children|panelVoice|default|hide|stopPropagation|enableExtruderVoice|show|unbind'.split('|'),0,{}))

inc/jquery.mb.flipText.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: mbicocchi@open-lab.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * jQuery.mb.components: jquery.mb.flipV
 * version: 1.1
 * © 2001 - 2009 Matteo Bicocchi (pupunzi), Open Lab
 *
 */

(function($) {
  var isIE=$.browser.msie;
  jQuery.fn.encHTML = function() {
    return this.each(function(){
      var me   = $(this);
      var html = me.text();
      me.text(html.replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;').replace(/'/g, escape("'")).replace(/"/g,escape('"')));
//      me.text(html.replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;').replace(/'/g, "’").replace(/"/g,"“"));
    });
  };
  $.mbflipText= {
    author:"Matteo Bicocchi",
    version:"1.1",
    flipText:function(tb){
      var UTF8encoded=$("meta[http-equiv=Content-Type]").attr("content") && $("meta[http-equiv=Content-Type]").attr("content").indexOf("utf-8")>-1;
      return this.each(function(){
        var el= $(this);
        var h="";
        var w="";

        var label="";
        var bgcol=(el.css("background-color") && el.css("background-color") != "rgba(0, 0, 0, 0)") ? el.css("background-color"):"#fff";
        var fontsize= parseInt(el.css('font-size'))>0?parseInt(el.css('font-size')):14;
        var fontfamily=el.css('font-family')?el.css('font-family').replace(/\'/g, '').replace(/"/g,''):"Arial";
        var fontcolor=el.css('color')? el.css('color'):"#000";

        if ($.browser.msie){
          if(!tb) el.css({'writing-mode': 'tb-rl', height:h, filter: 'fliph() flipv("") ', whiteSpace:"nowrap", lineHeight:fontsize+2+"px"}).css('font-weight', 'normal');
          label=$("<span style='writing-mode: tb-rl; whiteSpace:nowrap; height:"+h+"; width:"+w+"; line-height:"+(fontsize+2)+"px'>"+el.html()+"</span>");
        }else{

          var dim=el.getFlipTextDim(false);

          h=dim[1];
          w=dim[0];
          if(!isIE ) el.encHTML();
          var txt= el.text();

          var rot="-90";
          var ta="end";
          var xFix=0;
          var yFix=$.browser.opera ? parseInt(w)-(parseInt(w)/4): $.browser.safari?5:0;
          if (tb){
            yFix=$.browser.opera?20:0;
            xFix= $.browser.safari?(fontsize/4):0;
            rot="90, "+((parseInt(w)/2)-xFix)+", "+parseInt(w)/2;
            ta="start";
          }
          var onClick= el.attr("onclick") || el.attr("href");
          var clickScript= onClick?"<div class='pointer' style='position:absolute;top:0;left:0;width:100%;height:100%;background:transparent'/>":"";

          label=$("<object class='flip_label' style='height:"+h+"px; width:"+w+"px;' type='image/svg+xml' data='data:image/svg+xml; charset=utf-8 ," +
                  "<svg xmlns=\"http://www.w3.org/2000/svg\">" +
                  "<rect x=\"0\" y=\"0\" width=\""+w+"px\" height=\""+h+"px\" fill=\""+bgcol+"\" stroke=\"none\"/>"+
                  "<text  x=\"-"+xFix+"\" y=\""+yFix+"\" font-family=\""+fontfamily+"\"  fill=\""+fontcolor+"\" font-size=\""+fontsize+"\"  style=\"text-anchor: "+ta+"; " +
                  "dominant-baseline: hanging\" transform=\"rotate("+rot+")\" text-rendering=\"optimizeSpeed\">"+txt+"</text></svg>'></object>" +
                  clickScript +
                  "");
        }
        var wrapper= onClick ? $("<div/>").css("position","relative"): $("");
        var cssPos= el.wrap(wrapper).css("position")!="absolute" || el.css("position")!="fixed"  ?"relative" : el.css("position");
        el.html(label).css({position:cssPos, width:w});
      });
    },
    getFlipTextDim:function(enc){
      var el= $(this);
//      if(!enc && !isIE) el.encHTML();
      var txt= el.html();
      var fontsize= parseInt(el.css('font-size'));
      var fontfamily=el.css('font-family').replace(/'/g, '').replace(/"/g,'');
      if (fontfamily==undefined) fontfamily="Arial";
      var placeHolder=$("<span/>").css({position:"absolute",top:-100, whiteSpace:"noWrap", fontSize:fontsize, fontFamily: fontfamily});
      placeHolder.text(txt);
      $("body").append(placeHolder);
      var h = (placeHolder.outerWidth()!=0?placeHolder.outerWidth():(16+txt.length*fontsize*.60));
      var w = (placeHolder.outerHeight()!=0?placeHolder.outerHeight()+5:50);
      placeHolder.remove();
      return [w,h];
    }
  };
  $.fn.mbFlipText=$.mbflipText.flipText;
  $.fn.getFlipTextDim=$.mbflipText.getFlipTextDim;

})(jQuery);

inc/mbExtruder.js
/*******************************************************************************
 jquery.mb.components
 Copyright (c) 2001-2010. Matteo Bicocchi (Pupunzi); Open lab srl, Firenze - Italy
 email: mbicocchi@open-lab.com
 site: http://pupunzi.com

 Licences: MIT, GPL
 http://www.opensource.org/licenses/mit-license.php
 http://www.gnu.org/licenses/gpl.html
 ******************************************************************************/

/*
 * Name:jquery.mb.extruder
 * Version: 2.2
 * dependencies: jquery.metadata.js, jquery.mb.flipText.js, jquery.hoverintent.js
 */


(function($) {
  document.extruder=new Object();
  document.extruder.left = 0;
  document.extruder.top = 0;
  document.extruder.bottom = 0;
  document.extruder.right = 0;
  document.extruder.idx=0;
  var isIE=$.browser.msie;

  $.mbExtruder= {
    author:"Matteo Bicocchi",
    version:"2.2",
    defaults:{
      width:350,
      positionFixed:true,
      sensibility:800,
      position:"top",
      accordionPanels:true,
      top:"auto",
      extruderOpacity:1,
      flapMargin:35,
      textOrientation:"bt", // or "tb" (top-bottom or bottom-top)
      onExtOpen:function(){},
      onExtContentLoad:function(){},
      onExtClose:function(){},
      hidePanelsOnClose:true,
      closeOnExternalClick:true,
      autoCloseTime:0,
      autoOpenTime:0,
      slideTimer:300
    },

    buildMbExtruder: function(options){
      return this.each (function (){
        this.options = {};
        $.extend (this.options, $.mbExtruder.defaults);
        $.extend (this.options, options);
        this.idx=document.extruder.idx;
        document.extruder.idx++;
        var extruder,extruderContent,wrapper,extruderStyle,wrapperStyle,txt,closeTimer,openTimer;
        extruder= $(this);
        extruderContent=extruder.html();

        extruder.css("zIndex",100);

        var isVertical = this.options.position=="left" || this.options.position=="right";

        var extW= isVertical?1: this.options.width;

        var c= $("<div/>").addClass("content").css({overflow:"hidden", width:extW});
        c.append(extruderContent);
        extruder.html(c);

        var position=this.options.positionFixed?"fixed":"absolute";
        extruder.addClass("extruder");
        extruder.addClass(this.options.position);
        var isHorizontal = this.options.position=="top" || this.options.position=="bottom";
        extruderStyle=
                this.options.position=="top"?
                {position:position,top:0,left:"50%",marginLeft:-this.options.width/2,width:this.options.width}:
                        this.options.position=="bottom"?
                        {position:position,bottom:0,left:"50%",marginLeft:-this.options.width/2,width:this.options.width}:
                                this.options.position=="left"?
                                {position:position,top:0,left:0,width:1}:
                                {position:position,top:0,right:0,width:1};
        extruder.css(extruderStyle);
        if(!isIE) extruder.css({opacity:this.options.extruderOpacity});
        extruder.wrapInner("<div class='ext_wrapper'></div>");
        wrapper= extruder.find(".ext_wrapper");

        wrapperStyle={position:"absolute", width:isVertical?1:this.options.width};
        wrapper.css(wrapperStyle);


        if (isHorizontal){
          this.options.position=="top"?document.extruder.top++:document.extruder.bottom++;
          if (document.extruder.top>1 || document.extruder.bottom>1){
            alert("more than 1 mb.extruder on top or bottom is not supported jet... hope soon!");
            return;
          }
        }

        if ($.metadata){
          $.metadata.setType("class");
          if (extruder.metadata().title) extruder.attr("extTitle",extruder.metadata().title);
          if (extruder.metadata().url) extruder.attr("extUrl",extruder.metadata().url);
          if (extruder.metadata().data) extruder.attr("extData",extruder.metadata().data);
        }
        var flapFooter=$("<div class='footer'/>");
        var flap=$("<div class='flap'><span class='flapLabel'/></div>");
        if (document.extruder.bottom){
          wrapper.prepend(flapFooter);
          wrapper.prepend(flap);
        }else{
          wrapper.append(flapFooter);
          wrapper.append(flap);
        }

        txt=extruder.attr("extTitle")?extruder.attr("extTitle"): "";
        var flapLabel = extruder.find(".flapLabel");
        flapLabel.text(txt);
        if(isVertical){
          flapLabel.html(txt).css({whiteSpace:"noWrap"});//,height:this.options.flapDim
          var orientation= this.options.textOrientation == "tb";
          var labelH=extruder.find('.flapLabel').getFlipTextDim()[1];
          extruder.find('.flapLabel').mbFlipText(orientation);
        }else{
          flapLabel.html(txt).css({whiteSpace:"noWrap"});
        }

        if (extruder.attr("extUrl")){
          extruder.setMbExtruderContent({
            url:extruder.attr("extUrl"),
            data:extruder.attr("extData"),
            callback: function(){
              if (extruder.get(0).options.onExtContentLoad) extruder.get(0).options.onExtContentLoad();
            }
          })
        }else{
          var container=$("<div>").addClass("text").css({width:extruder.get(0).options.width-20, height:extruder.height()-20, overflowY:"auto"});
          c.wrapInner(container);
          extruder.setExtruderVoicesAction();
        }

        flap.bind("click",function(){
          if (!extruder.attr("open")){
            extruder.openMbExtruder();
          }else{
            extruder.closeMbExtruder();
          }
        }).bind("mouseenter",function(){
          if(extruder.get(0).options.autoOpenTime>0){
              openTimer=setTimeout(function(){
                extruder.openMbExtruder();
                $(document).one("click.extruder"+extruder.get(0).idx,function(){extruder.closeMbExtruder();});
              },extruder.get(0).options.autoOpenTime);
          }
        }).bind("mouseleave",function(){
          clearTimeout(openTimer);
        });

        c.bind("mouseleave", function(){
          if(extruder.get(0).options.closeOnExternalClick)
            $(document).one("click.extruder"+extruder.get(0).idx,function(){extruder.closeMbExtruder();});
          closeTimer=setTimeout(function(){

            if(extruder.get(0).options.autoCloseTime > 0){
              extruder.closeMbExtruder();
            }
          },extruder.get(0).options.autoCloseTime);
        }).bind("mouseenter", function(){
          clearTimeout(closeTimer);
          $(document).unbind("click.extruder"+extruder.get(0).idx);
        });

        if (isVertical){
          c.css({ height:"100%"});
          if(this.options.top=="auto") {
            flap.css({top:100+(this.options.position=="left"?document.extruder.left:document.extruder.right)});
            this.options.position=="left"?document.extruder.left+=labelH+this.options.flapMargin:document.extruder.right+= labelH+this.options.flapMargin;
          }else{
            flap.css({top:this.options.top});
          }
          var clicDiv=$("<div/>").css({position:"absolute",top:0,left:0,width:"100%",height:"100%",background:"transparent"});
          flap.append(clicDiv);
        }
      });
    },

    setMbExtruderContent: function(options){
      this.options = {
        url:false,
        data:"",
        callback:function(){}
      };
      $.extend (this.options, options);
      if (!this.options.url || this.options.url.length==0){
        alert("internal error: no URL to call");
        return;
      }
      var url=this.options.url;
      var data=this.options.data;
      var where=$(this), voice;
      var cb= this.options.callback;
      var container=$("<div>").addClass("container");
      if (!($.browser.msie && $.browser.version<=7))
        container.css({width:$(this).get(0).options.width});
      where.find(".content").wrapInner(container);
      $.ajax({
        type: "POST",
        url: url,
        data: data,
        success: function(html){
          where.find(".container").append(html);
          voice=where.find(".voice");
          voice.hover(function(){$(this).addClass("hover");},function(){$(this).removeClass("hover");});
          where.setExtruderVoicesAction();
          if (cb) {
            setTimeout(function(){cb();},100);
          }
        }
      });
    },

    openMbExtruder:function(c){
      var extruder= $(this);
      extruder.attr("open",true);
      $(document).unbind("click.extruder"+extruder.get(0).idx);
      var opt= extruder.get(0).options;
      extruder.addClass("open");
      if(!isIE) extruder.css("opacity",1);
      var position= opt.position;
      extruder.mb_bringToFront();
      if (position=="top" || position=="bottom"){
        extruder.find('.content').slideDown( opt.slideTimer);
        if(opt.onExtOpen) opt.onExtOpen();
      }else{

        if(!isIE) $(this).css("opacity",1);
        extruder.find('.ext_wrapper').css({width:""});
        extruder.find('.content').css({overflowX:"hidden", display:"block"});
        extruder.find('.content').animate({ width: opt.width}, opt.slideTimer);
        if(opt.onExtOpen) opt.onExtOpen();
      }
      if (c) {
        setTimeout(function(){
          $(document).one("click.extruder"+extruder.get(0).idx,function(){extruder.closeMbExtruder();});
        },100);
      }
    },

    closeMbExtruder:function(){
      var extruder= $(this);
      extruder.removeAttr("open");
      var opt= extruder.get(0).options;
      extruder.removeClass("open");
      $(document).unbind("click.extruder"+extruder.get(0).idx);
      if(!isIE) extruder.css("opacity",opt.extruderOpacity);
      if(opt.hidePanelsOnClose) extruder.hidePanelsOnClose();
      if (opt.position=="top" || opt.position=="bottom"){
        extruder.find('.content').slideUp(opt.slideTimer);
        if(opt.onExtClose) opt.onExtClose();
      }else if (opt.position=="left" || opt.position=="right"){
        extruder.find('.content').css({overflow:"hidden"});
        extruder.find('.content').animate({ width: 1 }, opt.slideTimer,function(){
          extruder.find('.ext_wrapper').css({width:1});
          extruder.find('.content').css({overflow:"hidden",display:"none"});
          if(opt.onExtClose) opt.onExtClose();
        });
      }
    }
  };

  jQuery.fn.mb_bringToFront= function(){
    var zi=10;
    $('*').each(function() {
      if($(this).css("position")=="absolute" ||$(this).css("position")=="fixed"){
        var cur = parseInt($(this).css('zIndex'));
        zi = cur > zi ? parseInt($(this).css('zIndex')) : zi;
      }
    });
    $(this).css('zIndex',zi+=1);
    return zi;
  };

  /*
   * EXTRUDER CONTENT
   */

  $.fn.setExtruderVoicesAction=function(){
    var extruder=$(this);
    var opt=extruder.get(0).options;
    var voices= $(this).find(".voice");
    voices.each(function(){
      var voice=$(this);
      if ($.metadata){
        $.metadata.setType("class");
        if (voice.metadata().panel) voice.attr("panel",voice.metadata().panel);
        if (voice.metadata().data) voice.attr("data",voice.metadata().data);
        if (voice.metadata().disabled) voice.attr("setDisabled", voice.metadata().disabled);
      }

      if (voice.attr("setDisabled"))
        voice.disableExtruderVoice();

      if (voice.attr("panel") && voice.attr("panel")!="false"){
        voice.append("<span class='settingsBtn'/>");
        voice.find(".settingsBtn").css({opacity:.5});
        voice.find(".settingsBtn").hover(
                function(){
                  $(this).css({opacity:1});
                },
                function(){
                  $(this).not(".sel").css({opacity:.5});
                }).click(function(){
          if ($(this).parents().hasClass("sel")){
            if(opt.accordionPanels)
              extruder.hidePanelsOnClose();
            else
              $(this).closePanel();
            return;
          }

          if(opt.accordionPanels){
            extruder.find(".optionsPanel").slideUp(400,function(){$(this).remove();});
            voices.removeClass("sel");
            voices.find(".settingsBtn").removeClass("sel").css({opacity:.5});
          }
          var content=$("<div class='optionsPanel'></div>");
          voice.after(content);
          $.ajax({
            type: "POST",
            url: voice.attr("panel"),
            data: voice.attr("data"),
            success: function(html){
              var c= $(html);
              content.html(c);
              content.children().not(".text")
                      .addClass("panelVoice")
                      .click(function(){
                extruder.closeMbExtruder();
              });
              content.slideDown(400);
            }
          });
          voice.addClass("sel");
          voice.find(".settingsBtn").addClass("sel").css({opacity:1});
        });
      }

      if (voice.find("a").length==0 && voice.attr("panel")){
        voice.find(".label").not(".disabled").css("cursor","pointer").click(function(){
          voice.find(".settingsBtn").click();
        });
      }

      if ((!voice.attr("panel") ||voice.attr("panel")=="false" ) && (!voice.attr("setDisabled") || voice.attr("setDisabled")!="true")){
        voice.find(".label").click(function(){
          extruder.hidePanelsOnClose();
          extruder.closeMbExtruder();
        });
      }
    });
  };

  $.fn.disableExtruderVoice=function(){
    var voice=$(this);
    var label = voice.find(".label");
    voice.removeClass("sel");
    voice.next(".optionsPanel").slideUp(400,function(){$(this).remove();});
    voice.attr("setDisabled",true);
    label.css("opacity",.4);
    voice.hover(function(){$(this).removeClass("hover");},function(){$(this).removeClass("hover");});
    label.addClass("disabled").css("cursor","default");
    voice.find(".settingsBtn").hide();
    voice.bind("click",function(event){
      event.stopPropagation();
      return false;
    });
  };

  $.fn.enableExtruderVoice=function(){
    var voice=$(this);
    voice.attr("setDisabled",false);
    voice.find(".label").css("opacity",1);
    voice.find(".label").removeClass("disabled").css("cursor","pointer");
    voice.unbind("click");
    voice.find(".settingsBtn").show();
  };

  $.fn.hidePanelsOnClose=function(){
    var voices= $(this).find(".voice");
    $(this).find(".optionsPanel").slideUp(400,function(){$(this).remove();});
    voices.removeClass("sel");
    voices.find(".settingsBtn").removeClass("sel").css("opacity",.5);
  };

  $.fn.openPanel=function(){
    var voice=$(this).hasClass("voice") ? $(this) : $(this).find(".voice");
    voice.each(function(){
      if($(this).hasClass("sel")) return;
      $(this).find(".settingsBtn").click();
    })
  };

  $.fn.closePanel=function(){
    var voice=$(this).hasClass("voice") ? $(this) : $(this).parent(".voice");
    voice.next(".optionsPanel").slideUp(400,function(){$(this).remove();});
    voice.removeClass("sel");
    $(this).removeClass("sel").css("opacity",.5);
  };

  $.fn.buildMbExtruder=$.mbExtruder.buildMbExtruder;
  $.fn.setMbExtruderContent=$.mbExtruder.setMbExtruderContent;
  $.fn.closeMbExtruder=$.mbExtruder.closeMbExtruder;
  $.fn.openMbExtruder=$.mbExtruder.openMbExtruder;

})(jQuery);